trustless gateway spec: add probing path (identity cid). (#491)
ipfs/specs: src/http-gateways/path-gateway.md

### `500` Internal Server Error

A generic server error returned when it is not possible to return a better
one. An internal server error signals the general unavailability of the
gateway.

### `502` Bad Gateway

Error that indicates that a Gateway was not able to produce response for a
known reason: for example, in the case of
[recursive gateways](#recursive-vs-non-recursive-gateways), in the event of
failure to find any providers for requested data. `502` indicates that the
request can be retried and is not a permanent failure.

This error response SHOULD include
[`Retry-After`](#retry-after-response-header) HTTP header to indicate how long
the client should wait before retrying.

> Gateways MUST return `404` instead of `502` when the content is known to be
> unretrivable: for example, when the Gateway is
> [non-recursive](#recursive-vs-non-recursive-gateways) and the content is known
> to not be available.

Reviewer: nit: MUST is not realistic

```suggestion
Gateways SHOULD return `404` instead of `502` when the content is known to be
unretrievable: for example, when the Gateway is
[non-recursive](#recursive-vs-non-recursive-gateways) and the content is known
to not be available.
```

### `504` Gateway Timeout

Error that indicates that the Gateway was not able to produce response under
set time limits: for example, when gateway failed to retrieve data from a
remote provider. `504` indicates that the request can be retried and is not a
permanent failure.

There is no generic timeout, Gateway implementations SHOULD set timeouts based
on specific use cases.

This error response SHOULD include
[`Retry-After`](#retry-after-response-header) HTTP header to indicate how long
the client should wait before retrying.

> Gateways MUST return `404` instead of `504` when the content is known to be
> unretrivable: for example, when the Gateway is
> [non-recursive](#recursive-vs-non-recursive-gateways) and the content is known
> to not be available.

Reviewer: ```suggestion
Gateways SHOULD return `404` instead of `504` when the content is known to be
unretrievable: for example, when the Gateway is
[non-recursive](#recursive-vs-non-recursive-gateways) and the content is known
to not be available.
```

### `Retry-After` (response header)

Gateway SHOULD return this header with error responses such as [`429 Too Many Requests`](#429-too-many-requests), [`504 Gateway Timeout`](#504-gateway-timeout) or `503` (server maintainance).

The "Retry-After" header indicates how long the user agent ought to wait before making a follow-up request. It uses the following syntax:

```

> Retry-After: <http-date>
> // or
> Retry-After: <delay-seconds>

Reviewer: :thought_balloon:   Date is virtually never used in real world, most services (Nginx, AWS, Cloudflare, Caddy) default to using delay in seconds for simplicity and practicality. 

Putting it here introduces unnecessary complexity (similar to range requests that are for more than one range – they are in spec, but nothing implements them correctly enough for it to be used). 

Let's simplify

```suggestion
Retry-After: <delay-seconds>
```

  - The downside of this approach is that it will always be slower than
    skipping child block resolution.

## Recursive vs non-recursive gateways

A *recursive Gateway* is a gateway which generally attempts to fetch content
from a third party it does not control by triggering lookups and retrievals. A
recursive Gateway may not know in advance whether it can obtain and return a
piece of content as the availability of it is out of its control. It may also

> suggest that clients retry failed requests later via `502` and `504` respones

Reviewer: ```suggestion
suggest that clients retry failed requests later via `502` and `504` responses
```

status codes.

A *non-recursive Gateway* is gateway which accesses a known content-set and,

> under normal operation conditions, knows with certainity whether content

Reviewer: ```suggestion
under normal operation conditions, knows with certainty whether content
```
